pkg/cloudprovider/drift.go

```diff
@@ -39,58 +40,51 @@ import (
 )
 
 const (
-	NodeClassDrift  cloudprovider.DriftReason = "NodeClassDrift"
-	K8sVersionDrift cloudprovider.DriftReason = "K8sVersionDrift"
-	ImageDrift      cloudprovider.DriftReason = "ImageDrift"
-	SubnetDrift     cloudprovider.DriftReason = "SubnetDrift"
+	NodeClassDrift       cloudprovider.DriftReason = "NodeClassDrift"
+	K8sVersionDrift      cloudprovider.DriftReason = "K8sVersionDrift"
+	ImageDrift           cloudprovider.DriftReason = "ImageDrift"
+	SubnetDrift          cloudprovider.DriftReason = "SubnetDrift"
+	KubeletIdentityDrift cloudprovider.DriftReason = "KubeletIdentityDrift"
 
 	// TODO (charliedmcb): Use this const across code and test locations which are signaling/checking for "no drift"
 	NoDrift cloudprovider.DriftReason = ""
 )
 
 func (c *CloudProvider) isNodeClassDrifted(ctx context.Context, nodeClaim *karpv1.NodeClaim, nodeClass *v1alpha2.AKSNodeClass) (cloudprovider.DriftReason, error) {
-	// First check if the node class is statically staticFieldsDrifted to save on API calls.
-	if staticFieldsDrifted := c.areStaticFieldsDrifted(nodeClaim, nodeClass); staticFieldsDrifted != "" {
-		return staticFieldsDrifted, nil
+	checks := []func(ctx context.Context, nodeClaim *karpv1.NodeClaim, nodeClass *v1alpha2.AKSNodeClass) (cloudprovider.DriftReason, error){
```

Comment:
Is the specific serialization of these checks meaningful in terms of either optimizing for "least expensive mean computation time", or "functional correctness"? If not, should we add a TODO or issue to parallelize them?

Reply:
My understanding is that order matters here, and cheaper checks are checked first. @charliedmcb could answer this more completely though.

I _think_ that these checks are all served from controller-runtime cached clients, which means that the general level of expense here is not actually that high, because the data we're examining is already in memory, so this is just a bunch of in-memory comparisons. The only thing that isn't true for is the maintenance windows stuff that @charliedmcb just added I believe.

Reply:
The intention of ordering is to start with the easier to complete (more reliable, less costly etc), with short circuiting on any drift condition being true. 

I think this space is cheap enough, there isn't a need to parallelize them.

Also, I do lean towards returning a consistent ordering for the drift conditions, even if it's not necessary the exact same one (maybe an earlier condition is later returned, once 2 cases become true).

Reply:
Correcting this some.

The MW stuff is in disruption not drift so doesn't apply here. The risky (high cost?) calls here are GET VM and GET NIC sent to Azure. The rest are served from controller-runtime clients and so cached. GetVM and GetNIC are the last pieces of drift checked.

I think above means we don't want to parallelize. AWS also doesn't parallelize these checks AFAIK.

Reply:
The two checks, which I think would be more expensive, are the checks for `isImageVersionDrifted` and `isSubnetDrifted` which have calls to `instanceProvider.Get` for the VM and `instanceProvider.GetNic` for the Nic respectively. However, they are located at the end of each set of drift checks.